File: backend/api/Models/Card.cs
using api.DTO;
using System.Text.Json.Serialization;

namespace api.Models;

public class Card
{
    public int Id { get; set; }
    public string Question { get; set; }
    public string Answer { get; set; }
    public byte[]? QuestionImage { get; set; }
    public byte[]? AnswerImage { get; set; }
    public CardStats CardStats { get; set; } = new CardStats();
    public int DeckId { get; set; }
    [JsonIgnore]
    public Deck Deck { get; set; }

    public Card() {}
    
    public Card(CardAddRequest cardCreateRequest)
    {
        Question = cardCreateRequest.Question;
        Answer = cardCreateRequest.Answer;
        QuestionImage = Convert.FromBase64String(cardCreateRequest.QuestionImage ?? "");
        AnswerImage = Convert.FromBase64String(cardCreateRequest.AnswerImage ?? "");
    }

    public void UpdateLearningStats(CardStatsUpdateRequest request)
    {
        CardStats.Repetitions = request.Repetitions;
        CardStats.Interval = request.Interval;
        CardStats.EaseFactor = request.EaseFactor;
        CardStats.DueDate = DateTime.Now.AddDays(request.Interval);
    }

    public Card(CardCreateRequest request)
    {
        Question = request.Question;
        Answer = request.Answer;
        DeckId = request.DeckId;
    }

    public Card Update(CardUpdateRequest request)
    {
        Question = request.Question;
        Answer = request.Answer;
        DeckId = request.DeckId;
        return this;
    }
}

File: mobile/lib/models/card.dart
import 'package:mobile/dtos/card_dto.dart';
import 'package:mobile/dtos/card_stats_update_request.dart';
import 'package:mobile/models/card_stats.dart';

class Card {
  int id;
  String question;
  String answer;
  String? questionImage;
  String? answerImage;
  CardStats cardStats;

  Card({
    required this.id,
    required this.question,
    required this.answer,
    required this.questionImage,
    required this.answerImage,
    required this.cardStats,
  });

  factory Card.fromJson(Map<String, dynamic> json) => Card(
        id: json["id"],
        question: json["question"],
        answer: json["answer"],
        questionImage: json["questionImage"],
        answerImage: json["answerImage"],
        cardStats: CardStats.fromJson(json["cardStats"]),
      );

  Map<String, dynamic> toJson() => {
        "id": id,
        "question": question,
        "answer": answer,
        "questionImage": questionImage,
        "answerImage": answerImage,
        "cardStats": cardStats,
      };

  void updateCardStats(CardStatsUpdateRequest cardStatsUpdateRequest) {
    cardStats.easeFactor = cardStatsUpdateRequest.easeFactor;
    cardStats.repetitions = cardStatsUpdateRequest.repetitions;
    cardStats.interval = cardStatsUpdateRequest.interval;
  }

  void editCard(CardDto cardDto) {
    question = cardDto.question;
    answer = cardDto.answer;
    questionImage = cardDto.questionImage;
    answerImage = cardDto.answerImage;
  }
}
